### PR TITLE
Validate json-schema against the pipeline config

### PR DESCRIPTION
**Pull request by corneliusweig**
This implements a basic test to validate the json-schema against the Skaffold go pipeline config (see #1915).

In its current shape, this test is pretty useless, because validation is basically only run against empty pipeline configs which consist of `apiVersion` and `kind`.

The fundamental problem here is, that `gojsonschema` validates documents against a schema. However, we have two schemas here: the json-schema and the schema encoded in our pipeline config. Validating the equivalence of the two schemas is quite a different problem. So this raises the question: is this the right approach?

However, if this test code should become more useful, we could add some non-trivial `skaffold.yaml` files to be validated against the json-schema.

### CODE REVIEW

File: hack/schemas/main_test.go [hunk 1]
```diff
@@ -73,3 +77,92 @@ func TestGenerators(t *testing.T) {
...
+	t.Helper()
+	schemaPath := fmt.Sprintf("%s/docs/content/en/schemas/%s.json", root, apiVersion)
+	if _, err := os.Stat(schemaPath); err == nil {
+		if content, err := ioutil.ReadFile(schemaPath); err != nil {
+			t.Errorf("unable to read existing schema for version %s", apiVersion)
+		} else {
+			return string(content)
+		}
+	} else if os.IsNotExist(err) {
+		t.Skip("could not find schema at", schemaPath)
```
Comment (nkubala) [inline hack/schemas/main_test.go]: should this be an error? if we ever changed the directory structure of the docs then we'd probably want this test to fail loudly and not silently

File: hack/schemas/main_test.go [hunk 2]
```diff
@@ -73,3 +77,92 @@ func TestGenerators(t *testing.T) {
...
+	}
+	return ""
```
Comment (nkubala) [inline hack/schemas/main_test.go]: if this gets returned and passed to `gojsonschema.NewStringLoader()` and then validated, it will pass right? do we consider this a success case? or should we just error out

**Comment by corneliusweig** [comment]
@nkubala Yes, it should probably fail at the two locations you pointed out. However, the much more severe problem is that the tests hardly _ever_ fail. I tried to explain this in the PR summary:

> The fundamental problem here is, that gojsonschema validates documents against a schema. However, we have two schemas here: the json-schema and the schema encoded in our pipeline config. Validating the equivalence of the two schemas is quite a different problem. So this raises the question: is this the right approach?

So maybe I should give a proper example: the `Version.Factory()` (pkg/skaffold/schema/versions.go:67) just returns an empty pipeline config. All non-empty fields are `apiVersion` and `kind`. However all other fields are optional, so the json-schema validation will virtually always succeed.

To fix this, we would need to create a collection of (valid) pipeline configs for all schema versions. And it would be very difficult to make these examples exhaustive. In addition, this creates a lot of maintenance burden, especially when introducing new schemas.

Bottom line: is this the right approach after all?

CC @balopat because he raised the original issue.

**Comment by nkubala** [comment]
@corneliusweig thanks for the explanation, I clearly hadn't thought hard enough about the actual schemas we generate in the first place. this seems like a problem to me - the fact that we generate two distinct schemas seems like we're setting ourselves up for huge technical debt down the road.

also, the fact that the json validator passes all the time since the fields are optional makes me think we're wasting our time here. I can't think of a better approach to validating the schema though, so maybe we should rethink the way we generate them in the first place :man_shrugging: 

my gut feeling is to close this PR. WDYT?

**Comment by corneliusweig** [comment]
@nkubala I have the same feeling, that this PR should be closed without merging. Maybe we wait for some final thoughts from @balopat ?

**Comment by nkubala** [comment]
going to close this one now. we can revisit if we feel a strong need for this later down the line. @corneliusweig thanks so much for the investigation!
